src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Serialization/ObjectReader.cs

```diff
@@ -253,33 +260,50 @@ public object ReadValue()
         }
     }
 
-    public (char[] array, int length) ReadCharArray(Func<int, char[]> getArray)
+    public byte[] ReadByteArray()
     {
-        var kind = (TypeCode)ReadByte();
+        var (result, _) = ReadRawArray<byte>(static (reader, array, length) => reader.Read(array, 0, length));
+        return result;
+    }
+
+    public char[] ReadCharArray()
+    {
+        var (result, _) = ReadCharArray(getArray: null);
+        return result;
+    }
 
-        (var length, _) = ReadArrayLengthAndElementKind(kind);
+    public (char[] array, int length) ReadCharArray(Func<int, char[]>? getArray)
+        => ReadRawArray(static (reader, array, length) => reader.Read(array, 0, length), getArray);
+
+    public (T[] array, int length) ReadRawArray<T>(
+        Func<BinaryReader, T[], int, int> read,
+        Func<int, T[]>? getArray = null)
+    {
+        // Defer to caller provided getArray if provided.  Otherwise, we'll just allocate the array ourselves.
+        getArray ??= static length => length == 0 ? [] : new T[length];
+
+        var length = ReadArrayLength();
         var array = getArray(length);
 
-        var charsRead = _reader.Read(array, 0, length);
+        var charsRead = read(_reader, array, length);
 
         return (array, charsRead);
     }
 
     /// <summary>
     /// A reference-id to object map, that can share base data efficiently.
     /// </summary>
-    private readonly struct ReaderReferenceMap<T> : IDisposable
-        where T : class
+    private readonly struct ReaderReferenceMap : IDisposable
```

Comment:
>ReaderReferenceMap [](http://example.com/codeflow?start=28&length=18)

Should WriterReferenceMap be switched from object to string to match?

Reply:
done.